algorithms/feature-tracking/include/feature-tracking/grided-detector.h

```diff
@@ -0,0 +1,209 @@
+    KeyPointDataIterator nearest_corners_end =
+        keypoint_data_vector.begin() + corner_row_LUT[y_bottom];
+
+    for (KeyPointDataIterator it = nearest_corners_begin;
+         it != nearest_corners_end; ++it) {
+      if (it->keypoint_index == current_keypoint_data.keypoint_index ||
+          erase_keypoints[it->keypoint_index] ||
+          !IsInsideCircle(current_keypoint_data, it)) {
+        continue;
+      }
+      const float response_threshold =
+          ratio_threshold * current_keypoint_data.response;
+      if (response_threshold > it->response) {
+        erase_keypoints[it->keypoint_index] = true;
+      }
+    }
+  }
+
+  // Remove the flaged non-maximum keypoints.
+  std::vector<bool>::iterator it_erase = erase_keypoints.begin();
+
+  std::vector<cv::KeyPoint>::iterator it_erase_from = std::remove_if(
+      keypoints->begin(), keypoints->end(),
+      [&it_erase](const cv::KeyPoint & /*keypoint*/) -> bool {
+        return *it_erase++ == true;
+      });
+  keypoints->erase(it_erase_from, keypoints->end());
+}
+
+inline void detectKeypointsGrided(
+    const cv::Ptr<cv::FeatureDetector>& detector, const cv::Mat& image,
+    const cv::Mat& detection_mask, size_t max_total_keypoints,
+    float nonmaxsuppression_radius, float nonmaxsuppression_ratio_threshold,
+    size_t grid_rows, size_t grid_cols, std::vector<cv::KeyPoint>* keypoints) {
+  CHECK_NOTNULL(keypoints)->clear();
+  CHECK_GE(grid_rows, 1u);
+  CHECK_GE(grid_cols, 1u);
+
+  if (image.empty() || max_total_keypoints < grid_rows * grid_cols) {
+    keypoints->clear();
+    return;
+  }
+  keypoints->reserve(max_total_keypoints);
+
+  constexpr double kCellNumFeaturesScaler = 2.0;
+  int max_per_cell =
+      kCellNumFeaturesScaler * max_total_keypoints / (grid_rows * grid_cols);
+
+  std::mutex m_keypoints;
+  auto detectFeaturesOfGridCells = [&](const std::vector<size_t>& range) {
+    for (int cell_idx : range) {
+      int celly = cell_idx / grid_cols;
+      int cellx = cell_idx - celly * grid_cols;
+
+      cv::Range row_range(
+          (celly * image.rows) / grid_rows,
+          ((celly + 1) * image.rows) / grid_rows);
+      cv::Range col_range(
+          (cellx * image.cols) / grid_cols,
+          ((cellx + 1) * image.cols) / grid_cols);
+
+      cv::Mat sub_image = image(row_range, col_range);
+      cv::Mat sub_mask;
+      if (!detection_mask.empty()) {
+        sub_mask = detection_mask(row_range, col_range);
+      }
+
+      std::vector<cv::KeyPoint> sub_keypoints;
+      sub_keypoints.reserve(max_per_cell);
+
+      detector->detect(sub_image, sub_keypoints, sub_mask);
+
+      std::vector<cv::KeyPoint>::iterator it = sub_keypoints.begin(),
+                                          end = sub_keypoints.end();
+      for (; it != end; ++it) {
+        it->pt.x += col_range.start;
+        it->pt.y += row_range.start;
+      }
+
+      if (nonmaxsuppression_radius > 0.0) {
+        localNonMaximumSuppression(
```

Comment:
I would suggest nms after grided detection. Otherwise, you will not be able to suppresss keypoints that are both inside the radius but separated by the grid.